File: src/main/java/com/stolser/javatraining/webproject/controller/FrontController.java
package com.stolser.javatraining.webproject.controller;

import com.stolser.javatraining.webproject.controller.utils.HttpUtils;
import com.stolser.javatraining.webproject.view.ViewResolver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class FrontController extends HttpServlet {
    private static final Logger LOGGER = LoggerFactory.getLogger(FrontController.class);
    private static final String USER_ID_REQUEST_URI = "User id = {}. requestURI = {}";
    private static RequestProvider requestProvider = RequestProviderImpl.getInstance();

    @Override
    public void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        processRequest(request, response);
    }

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        processRequest(request, response);
    }

    private void processRequest(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String viewName;

        try {
            viewName = requestProvider.getRequestProcessor(request).getViewName(request, response);

        } catch (Exception e) {
            LOGGER.error(USER_ID_REQUEST_URI,
                    HttpUtils.getUserIdFromSession(request), request.getRequestURI(), e);

            viewName = ApplicationResources.getErrorViewName(e);
        }

        dispatch(viewName, request, response);
    }

    private void dispatch(String viewName, HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        if (viewName != null) {
            String page = ViewResolver.getPageByViewName(viewName);

//            System.out.println("forwarding to '" + page + "'");

            RequestDispatcher dispatcher = request.getRequestDispatcher(page);
            dispatcher.forward(request, response);
        }
    }
}

File: src/main/scala/com/stolser/javatraining/webproject/controller/FrontController.scala
package com.stolser.javatraining.webproject.controller

import java.io.IOException

import com.stolser.javatraining.webproject.controller.request.processor.{DispatchException, RequestProvider, RequestProviderImpl}
import com.stolser.javatraining.webproject.controller.utils.HttpUtils
import com.stolser.javatraining.webproject.view.{JspViewResolver, ViewResolver}
import javax.servlet.{RequestDispatcher, ServletException}
import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
import org.slf4j.{Logger, LoggerFactory}

/**
  * Created by Oleg Stoliarov on 10/13/18.
  * Implementation of the Front Controller pattern.
  */
class FrontController extends HttpServlet {
	private val LOGGER = LoggerFactory.getLogger(classOf[FrontController])
	private val DISPATCHING_TO_THE_VIEW_NAME = "Dispatching to the viewName = '%s'."
	private val INCORRECT_THE_DISPATCH_TYPE = "Incorrect the dispatch type of the abstractViewName: %s"
	private val requestProvider = RequestProviderImpl
	private val viewResolver = JspViewResolver.getInstance

	@throws[ServletException]
	@throws[IOException]
	override def doGet(request: HttpServletRequest, response: HttpServletResponse): Unit =
		processRequest(request, response)

	@throws[ServletException]
	@throws[IOException]
	override protected def doPost(request: HttpServletRequest, response: HttpServletResponse): Unit =
		processRequest(request, response)

	@throws[ServletException]
	@throws[IOException]
	private def processRequest(request: HttpServletRequest, response: HttpServletResponse): Unit = {
		try {
			val abstractViewName = requestProvider.getRequestProcessor(request).process(request, response)
			dispatch(abstractViewName, request, response)
		} catch {
			case e: RuntimeException =>
				logExceptionAndRedirectToErrorPage(request, response, e)
		}
	}

	private def dispatch(abstractViewName: String, request: HttpServletRequest, response: HttpServletResponse): Unit = {
		val viewNameParts = abstractViewName.split(":")
		val dispatchType = viewNameParts(0)
		val viewName = viewNameParts(1)
		dispatchType match {
			case "forward" =>
				sendForward(request, response, viewName)
			case "redirect" =>
				HttpUtils.sendRedirect(request, response, viewName)
			case "noAction" => ()
			case _ =>
				throw new IllegalArgumentException(String.format(INCORRECT_THE_DISPATCH_TYPE, abstractViewName))
		}
	}

	def sendForward(request: HttpServletRequest, response: HttpServletResponse, viewName: String): Unit = {
		try {
			val dispatcher = request.getRequestDispatcher(viewResolver.resolvePrivateViewName(viewName))
			dispatcher.forward(request, response)
		} catch {
			case e @ (_: ServletException | _: IOException) =>
				throw new DispatchException(String.format(DISPATCHING_TO_THE_VIEW_NAME, viewName), e)
		}
	}

	private def logExceptionAndRedirectToErrorPage(request: HttpServletRequest,
												   response: HttpServletResponse,
												   e: RuntimeException): Unit = {
		LOGGER.error(s"User id = ${HttpUtils.getUserIdFromSession(request)}. requestURI = ${request.getRequestURI}", e)
		HttpUtils.sendRedirect(request, response, viewResolver.resolvePublicViewName(HttpUtils.getErrorViewName(e)))
	}
}
